parteh/PRTGenericMod.F90

```diff
@@ -0,0 +1,1196 @@
+              write(fates_log(),*) ' pft id: ',ipft
+              write(fates_log(),*) ' position id: ',position_id
+              write(fates_log(),*) ' organ id: ',prt_instance%state_descriptor(i_var)%organ_id
+              write(fates_log(),*) ' species_id: ',prt_instance%state_descriptor(i_var)%spec_id
+              write(fates_log(),*) ' position id: ',i_pos
+              write(fates_log(),*) ' symbol: ',trim(prt_instance%state_descriptor(i_var)%symbol)
+              write(fates_log(),*) ' longname: ',trim(prt_instance%state_descriptor(i_var)%longname)
+              write(fates_log(),*) ' err: ',err,' max error: ',calloc_abs_error
+              write(fates_log(),*) ' terms: ', this%variables(i_var)%val(i_pos), &
+                                               this%variables(i_var)%val0(i_pos), &
+                                               this%variables(i_var)%net_art(i_pos), &
+                                               this%variables(i_var)%turnover(i_pos), &
+                                               this%variables(i_var)%burned(i_pos)
+              write(fates_log(),*) ' Exiting.'
+              call endrun(msg=errMsg(__FILE__, __LINE__))
+           end if
+
+        end do
+     end do
+
+     return
+   end subroutine CheckMassConservation
+
+   ! ====================================================================================
+   
+   function GetState(this, organ_id, species_id, position_id) result(sp_organ_val)
+      
+      ! This function returns the current amount of mass for
+      ! any combination of organ and species.  If a position
+      ! is provided, it will use it, but otherwise, it will sum over
+      ! all dimensions.  It also can accomodate all_carbon_species, which
+      ! will return the mass of all carbon isotopes combined.
+
+      class(prt_vartypes)   :: this
+      integer,intent(in)    :: organ_id
+      integer,intent(in)    :: species_id
+      integer,intent(in),optional :: position_id
+      real(r8)              :: sp_organ_val
+
+      integer :: i_pos
+      integer :: ispec
+      integer :: num_species
+      integer,dimension(max_spec_per_group) :: spec_ids 
+      integer :: i_var
+      
+      sp_organ_val = 0.0_r8
+      
+      if(species_id == all_carbon_species) then
+         spec_ids(1:3) = carbon_species(1:3)
+         num_species  = 3
+      else
+         num_species  = 1
+         spec_ids(1) = species_id
+      end if
+
+      if(present(position_id)) then
+         i_pos = position_id
+      
+         do ispec = 1,num_species
+            i_var = prt_instance%sp_organ_map(organ_id,spec_ids(ispec))
+            if(i_var>0) sp_organ_val = sp_organ_val + this%variables(i_var)%val(i_pos)
+         end do
+
+      else
+         
+         do ispec = 1,num_species
+            
+            i_var = prt_instance%sp_organ_map(organ_id,spec_ids(ispec))
+            if(i_var>0)then
+
+               do i_pos = 1, prt_instance%state_descriptor(i_var)%num_pos
```

Comment:
if prt_instance%state_descriptor(i_var)%num_pos is >1, then does this overwrite sp_organ_val as it goes through the loop (thus always ending up with the last value?